File: backend/app/schemas/evaluation.py
"""评估模块的请求/响应 Schema。"""
from datetime import datetime

from pydantic import BaseModel, Field


# --- 文档列表 ---
class SampleDocOut(BaseModel):
    """可选的已入库文档信息。"""
    id: str
    filename: str
    size_bytes: int
    file_type: str


# --- 出题 ---
class GenerateQuestionsReq(BaseModel):
    """出题请求：指定文档和每文档题目数。"""
    doc_ids: list[str] = Field(..., min_length=1, description="选中的文档ID列表")
    num_per_doc: int = Field(3, ge=1, le=10, description="每个文档生成的题目数")


class QuestionItem(BaseModel):
    """单道测试题。"""
    question: str
    ground_truth: str


# --- 评估任务 ---
class StartEvalReq(BaseModel):
    """启动评估请求：传入题目列表和源文档。"""
    questions: list[QuestionItem] = Field(..., min_length=1)
    source_files: list[str] = Field(default_factory=list)


class EvalTaskOut(BaseModel):
    """评估任务状态响应。"""
    id: str
    status: str
    progress: int
    total_questions: int
    current_step: str
    avg_faithfulness: float | None = None
    avg_answer_relevancy: float | None = None
    avg_context_precision: float | None = None
    avg_context_recall: float | None = None
    source_files: str  # JSON 字符串
    error: str | None = None
    # LLM 生成的整体评估点评(最低指标+优化建议);评估完成后写入,可为空
    review: str | None = None
    created_at: datetime
    finished_at: datetime | None = None

    model_config = {"from_attributes": True}


class EvalRecordOut(BaseModel):
    """单条评估明细响应。"""
    id: str
    question: str
    ground_truth: str
    response: str
    retrieved_contexts: str  # JSON 字符串
    faithfulness: float | None = None
    answer_relevancy: float | None = None
    context_precision: float | None = None
    context_recall: float | None = None

    model_config = {"from_attributes": True}

File: frontend/src/api/evaluation.ts
/**
 * RAG 评估模块 API 调用封装。
 */
import client from './client'

export interface SampleDoc {
  name: string
  size_bytes: number
  file_type: string
}

export interface QuestionItem {
  question: string
  ground_truth: string
}

export interface EvalTask {
  id: string
  status: string
  progress: number
  total_questions: number
  current_step: string
  avg_faithfulness: number | null
  avg_answer_relevancy: number | null
  avg_context_precision: number | null
  avg_context_recall: number | null
  source_files: string
  error: string | null
  created_at: string
  finished_at: string | null
}

export interface EvalRecord {
  id: string
  question: string
  ground_truth: string
  response: string
  retrieved_contexts: string
  faithfulness: number | null
  answer_relevancy: number | null
  context_precision: number | null
  context_recall: number | null
}

/** 获取可选文档列表 */
export const getSampleDocs = () => client.get<SampleDoc[]>('/evaluation/docs')

/** 根据选中文档生成测试题 */
export const generateQuestions = (file_names: string[], num_per_doc = 3) =>
  client.post<QuestionItem[]>('/evaluation/generate', { file_names, num_per_doc })

/** 启动评估任务 */
export const startEvaluation = (questions: QuestionItem[], source_files: string[]) =>
  client.post<EvalTask>('/evaluation/start', { questions, source_files })

/** 获取历史评估任务列表 */
export const listEvalTasks = () => client.get<EvalTask[]>('/evaluation/tasks')

/** 获取单个任务状态(轮询用) */
export const getEvalTask = (taskId: string) => client.get<EvalTask>(`/evaluation/tasks/${taskId}`)

/** 获取评估明细结果 */
export const getEvalRecords = (taskId: string) =>
  client.get<EvalRecord[]>(`/evaluation/tasks/${taskId}/records`)
